src/OpenLoco/src/Ui/ToolManager.cpp

```diff
@@ -141,23 +170,198 @@ namespace OpenLoco::ToolManager
         _toolWindowType = toolWindowType;
     }
 
-    // 0x00523393
     Ui::CursorId getToolCursor()
     {
         return _toolWindowCursor;
     }
+
     void setToolCursor(Ui::CursorId toolWindowCursor)
     {
         _toolWindowCursor = toolWindowCursor;
     }
 
-    // 0x00523394
     int16_t getToolWidgetIndex()
     {
         return _toolWidgetIndex;
     }
-    void setToolWidgetIndex(uint16_t toolWidgetIndex)
+
+    void setToolWidgetIndex(int16_t widget)
+    {
+        _toolWidgetIndex = widget;
+    }
+
+
+    CursorId getCursor(bool& out, int16_t x, int16_t y)
+    {
+        auto w = WindowManager::find(currentTool.type, currentTool.window);
+        if (w == nullptr)
+        {
+            return currentTool.cursor;
+        }
+        return currentTool.getCursor(*w, out, x, y);
+    }
+
+    bool fireEvent(ToolEventType event, int16_t x, int16_t y, int16_t mouseWheel)
+    {
+        auto w = WindowManager::find(currentTool.type, currentTool.window);
+        if (w == nullptr)
+        {
+            return false;
+        }
+        return currentTool.fireEvent(event, x, y, mouseWheel);
+    }
+
+    bool ToolBase::fireEvent(ToolEventType event, int16_t x, int16_t y, int16_t mouseWheel)
     {
-        _toolWidgetIndex = toolWidgetIndex;
+        if (!Input::hasFlag(Input::Flags::toolActive) && event != ToolEventType::onStop)
+        {
+            return false;
+        }
+
+        input.set(x, y, mouseWheel);
+
+        switch (event)
+        {
+            case ToolEventType::onMouseDrag:
+                input.dragging = true;
+                break;
+            case ToolEventType::onMouseDragEnd:
+            case ToolEventType::onStart:
+                input.dragging = false;
+                break;
+            default:
+                break;
+        }
+
+        if (!hasEvent(event))
+        {
+            return false;
+        }
+
+        auto w = WindowManager::find(type, window);
+        if (w == nullptr)
+        {
+            return false;
+        }
+        switch (event)
+        {
+            case ToolEventType::onStart:
+                onStart(*w, event);
```

Comment:
This is great except it calls the base class methods not the derived class methods. Is it possible? Nobody seems to know.

Reply:
I've set the base methods =0 which makes it an abstract class, which is fine except that it requires derived classes to override all virtual methods, which adds overhead. What would be the best way to handle that?

Reply:
> Is it possible to call a derived class method from a base class? Nobody seems to know.

This is just basic virtual functions; am I missing something? In general:

```
class Base {
public:
    virtual void doSomething() {    // Virtual function in the base class
        std::cout << "Base class doing something\n";
    }

    void callDoSomething() {
        doSomething(); // Calls the most derived version of doSomething
    }
};

class Derived : public Base {
public:
    void doSomething() override { // Override the virtual function
        std::cout << "Derived class doing something\n";
    }
};
```

even if you have a `Base* ptr = new Derived();`, `ptr->callDoSomething();` will return "Derived..."

Reply:
When I use the VS debugger, stepping into the calls here goes to the base class methods, not the derived class.

Reply:
The issue was using refs not pointers.